File: sprint6/C++/lesson6code1.cpp
#include <stack>
#include <vector>

std::stack<int> order; // В этом стеке будет записан порядок обхода.
std::vector<int> color(n, white);

void topSort(int v) {
    color[v] = gray;
    for (int w : adjList[v]) {
        if (color[w] == white) {
            topSort(w);
        }
    }
    color[v] = black;
    order.push(v); // Кладём обработанную вершину в стек.
}

void mainTopSort() {
    for (int i = 0; i < n; i++) {
        if (color[i] == white) {
            topSort(i);
        }
    }
}

File: sprint6/By languages/Java/lesson6code1.java
Stack<Integer> order = new Stack<>();
int[] color = new int[n];

void topSort(int v) {
    color[v] = 1;
    for (int w : adjList[v]) {
        if (color[w] == 0) {
            topSort(w);
        }
    }
    color[v] = 2;
    order.push(v);
}

void mainTopSort() {
    for (int i = 0; i < n; i++) {
        if (color[i] == 0) {
            topSort(i);
        }
    }
}
